### DISCUSSION TITLE
Citations = K vectors?

### DISCUSSION THREAD

**Discussion by Antoine101** [open-webui/open-webui#17287] [Q&A] [answered]
Hi everyone,

When querying a knowledge base, the returned response includes links to referenced documents (inline and in the footing). When clicking on these links, we can see citations that were considered in these documents, with the source (name of the document), the relevance score and the content.

For one of my queries, made on a knowledge base of two documents only, the two documents are referenced and I get 4 citations in the first one and three in the second one.

I understand (but correct me if I'm wrong) that citations correspond to vectors from these documents. What I don't understand however is why I have 7 citations for one query, when I set a Top K of my reranker to 3 (50 initially passed to reranker and then the reranker returns the best 3). I would expect the answer to my query to only be based on 3 vectors, thus 3 citations, and not 4+3=7.

Can someone please clarify this point for me please?
<img width="1436" height="723" alt="Capture d’écran 2025-09-08 à 16 33 11" src="https://github.com/user-attachments/assets/d3f481aa-20ea-4bed-b3b9-4b06de7cc60c" />
<img width="1436" height="723" alt="Capture d’écran 2025-09-08 à 16 33 25" src="https://github.com/user-attachments/assets/28ca1324-d8eb-4121-8266-dffbd6f4ae4e" />

**Comment by mahenning** [comment]
I understand this is confusing, let me explain the RAG procedure (with e.g. k=5, k_reranker=3):
- the user asks aquestion about a knowledge base (or a document in chat)
- the last chat messages are used to generate 1-3 queries (most of the time 3 queries) that are used to extract knowledge
- _per query_, the k=5 best vectors are searched, then reranked and the best k_reranker=3 results are taken
- then the results from the queries are merged. When the same chunk is returned from more than one query, duplicates are removed
- the chat LLM gets the resulting chunks together with their relevance scores (relevance score is computed against the query where the chunk came from)

This results in 3-9 chunks returned. In your case it seems 2 chunks had duplicates, so you got only 7.

I know this is usually not what the user expects when they set a certain k/k_reranker value. The problem is that the generated queries could be very different and the chunk relevance is computed in comparison to the current query. Possible solutions:
- limit RAG to 1 query. This can lead to less knowledge extracted, because sometimes multiple quesions have to be asked to get a good answer
- Hard cap the merged results from the queries to k/k_reranker. This could be done, but the relevance score from the chunks are computed against the query that extracted them. For example:
  - query 1 asks a very specific thing about a smaller topic of the chat question, gets 3 specific chunks with high relevance back
  - query 2 is a more general question, that returns 3 chunks with less relevance score, that are maybe important for the general understanding
  -> In this scenario, the chunks from query 1 would "win" due to high relevance scores and only they would be returned.
Both cases would correctly limit the results the chat LLM get to k/k_reranker, but the quality of the information retrieval could get worse. I didn't find a good solution for that yet.

Hope I could help

**Comment by Antoine101** [comment]
Thank you for taking the time to reply @mahenning !

I don't understand your second sentence "the last chat messages are used to generate 1-3 queries (most of the time 3 queries) that are used to extract knowledge", which makes it difficult for me to understand some of the following points you raise.

What are those multiple queries? Where are they coming from? How are they generated?

Would you be able to provide a quick example to illustrate your point please?

**Comment by mahenning** [comment, marked as answer]
Sure thing. You can see the default RAG template that will generate the queries here (`DEFAULT_QUERY_GENERATION_PROMPT_TEMPLATE`): https://docs.openwebui.com/getting-started/env-configuration/#query_generation_prompt_template
The 
```
### Chat History:
<chat_history>
{{MESSAGES:END:6}}
</chat_history>
```
part is filled out on the fly with the last 6 messages from the chat.

So, lets say the user is already a few messages into a topic about "python vs R" , and then uploads a document and writes 
> "Can this help? Can you tell me more about it?"

Using that prompt directly as a query for RAG woudn't be good. The user indicates a topic with this question (python vs R), but on its own "Can this help? Can you tell me more about it?" means nothing really. So, together with this question, and the context of the chat, queries are generated that look maybe like:
```
'Python vs R for data science',
'When to use Python or R for statistical analysis',
'Comparison of Python and R libraries for data analysis'
```
(I took this directly from a test chat and extracted the generated queries)
With a query like this, finding sematic relevant information in (potentially huge) document data is much easier. These queries are then used to find relevant text chunks in the documents.

One could argue that the queries are a bit same-y, but this should be more or less fine as queries that are very close together tend to retrieve the same chunks, which are filtered out in the duplicate step.

**Reply by Antoine101** [reply]
Thank you so much for this detailed answer! I get it now.
I will try to see what queries are generated for my use case by looking at LiteLLM logs.

**Reply by mahenning** [reply]
<del>Unfortunately the queries are never logged. I got them when I halted the execution with a debugger and printed out the variable.</del>
Edit: See comment below.

**Reply by Antoine101** [reply]
But LiteLLM is just acting as a router and the logs show what is being sent to the LLM and what is being outputed so it should be there no?

**Reply by mahenning** [reply]
Ah of course, my bad! Then it should be possible. The query generation is "just" another LLM call.

**Reply by runyournode** [reply]
Thank you @mahenning for the answer.

Would you also know:
 - if using N knowledge bases in a rag model could possibly generate k_rerank chunks per kb ? Or are all theses chunks aggregated and filtered after the reranking process ? (in other word, if I have k_rerank, N kb and 3 queries, ignoring dupplicated chunks, could k_rerank * N * 3 chunks passed to the llm in the final stage ?)
 - how is BM25 implemented in owui ? in particular:
   - does it retrieve its own chunks or does it only score the already retrieved RAG_TOP_K <* num_queries> chunks ? 
   - when using an external reranker, is BM25 still being used internally (or is it expected to be implemented in the external reranker ?)
   - how are pertinence score normalized [0-1] in owui ? When using the same reranker model externally, I get pertinence score as low as -10, whereas when using it internally it is always bound in [0,1]
   
  Thank you :-)

**Reply by mahenning** [reply]
- It seems that for each element (collection, single file..) a search is started with all queries. That means, if you add 2 collections and a single file and asks something:
  - up to 3 queries are generated
  - for each of these 3 sources (2 collections, 1 file), all 3 queries are used
  - For each source: 
    - Each query can output k_reranker results
    - These results are then merged together (remove duplicated results from different queries) and up to k (not k_reranker) aggregated results are returned

So yes, it seems the maximum (no duplicates) are k_reranker * num_queries * sources, given that k > 3* k_reranker. I'm not sure why k is used as a maximum when merging / filtering duplicates. Personally, I'd just use all sources together for each query, and also maybe limit even the results of all 3 queries to stay  in the `k_reranker` (or k without hybrid search) limit.
 
- BM25 retrieves its own separate k chunks. VectorRetriever also retrieves its own k chunks. They then get ranked and fused, meaning the actual scores of BM25 and the VectorRetriever are ignored, only the _rank_ (e.g. 3rd place out of 10) counts. If a chunk is present both in the results of BM25 and VectorRetriever, they get a higher fused rank. This is done because BM25 and VectorRetriever use different scores which are not comparable. The ranked documents are then reranked and scored by the Reranker.
  - BM25 is used internally, it doesnt run on the GPU 

> how are pertinence score normalized [0-1] in owui ? When using the same reranker model externally, I get pertinence score as low as -10, whereas when using it internally it is always bound in [0,1]

I'm not exactly sure why it is not bound in [0, 1] for you externally. I saw that the activation function for the loaded reranker was sigmoid, which explains the [0, 1] output. Do you maybe not load your reranker as a CrossEncoder?

**Reply by runyournode** [reply]
Thank you for the answer. 
>  I'm not sure why k is used as a maximum when merging / filtering duplicates.
 - Can you confirm the `k` you are referring, is (ignoring duplicates)  `k = k_rerank * num_queries * sources` and not the `RAG_TOP_K` we can set in env ?
>"BM25 retrieves its own separate k chunks. VectorRetriever also retrieves its own k chunks. They then get ranked and fused ... The ranked documents are then reranked and scored by the Reranker"
-  I understand that [rff](https://cormack.uwaterloo.ca/cormacksigir09-rrf.pdf) is used in langchain to fuse and rank the two lists (len=`RAG_TOP_K`) into a single fused_list (len=2*`RAG_TOP_K`). Then, do we pass to the reranker only the first `RAG_TOP_K` item of the  fused_list ( *i.e.* fused_list[:`RAG_TOP_K`]) ? If not, then why bother with a rff ranking ?

>   Do you maybe not load your reranker as a CrossEncoder?
 - You are perfectly right ! I was loading with `model = AutoModelForSequenceClassification.from_pretrained("BAAI/bge-reranker-v2-m3")` and then `model([query]*n,[n_chunks]).logits` without any activation function. Thanks !

**Reply by mahenning** [reply]
Regarding the k parameter, it is `RAG_TOP_K`. You can see it here in the code:
https://github.com/open-webui/open-webui/blob/171021cfa4276f63fd9fd7f31fa0c904fb13c24c/backend/open_webui/retrieval/utils.py#L403
The function for that return does get `k` and `k_reranker`:
https://github.com/open-webui/open-webui/blob/171021cfa4276f63fd9fd7f31fa0c904fb13c24c/backend/open_webui/retrieval/utils.py#L330-L339

So the `merge_and_sort_query_results` gets the `k_reranker` * num_queries results, filters duplicates and returns the best `k`.

> Then, do we pass to the reranker only the first RAG_TOP_K item of the fused_list ( i.e. fused_list[:RAG_TOP_K]) ? If not, then why bother with a rff ranking ?

As far as i saw it, the reranker gets the full fused list. So yes, rank fusing could be useless, but I think it ist not a great bottleneck. Even given k=1000, rank-fusing 2 1000 item lists should be quick in comparison to search a vector space and invoking the reranker. At least it fuses/discards doubled entries.

**Comment by mahenning** [comment]
@Antoine101 In the current version of OWUI you can actually see the queries when you click on the "Retrieved x sources" at the top of the LLM answer. I'm not sure when that was added.
<img width="666" height="174" alt="grafik" src="https://github.com/user-attachments/assets/011a6032-67f5-488d-b4e3-8915f703db36" />

**Reply by Antoine101** [reply]
Thanks for the heads-up @mahenning ! I couldn't see anything on v0.6.26 so I pulled v0.6.28 (the latest) and now I can see the same thing as you, the three queries generated from my initial query and past chat messages.

Do you think what you are highlighting is passed as is to the LLM or are these only summaries of the generated queries?

**Reply by Classic298** [reply]
was added in 0.6.27 as part of the websearch design refactor (see changelog). Now all queries are shown, RAG queries and websearch queries for transparency :)

**Reply by mahenning** [reply]
These are the original queries used for the RAG (and also web search it seems).
Thanks for the information @Classic298!
